### PR TITLE
feat: sales uom and purchase uom  in stock ledger

### PR DESCRIPTION
**Pull request by Dilshanavs**
A new filter "Select UOM" added in Stock Ledger Report. In this filter we can select Default UOM, Sales UOM and Purchase UOM..
Result of selecting 'Default UOM':
image

Sales UOM:
image

Purchase UOM:
image

`no-doc`

### CODE REVIEW

**Comment by bosue** [comment]
Looks like this is an alternative implementation of something like #37332? Reusing the already existing UoMs (default, purchase, sales) seems quite a bit leaner and cleaner than adding yet another one.

But we all will have to figure out the details. I have to admit it’s a bit hard to grasp. Unsure if any of these is currently possible, but if yes, we might want to either use two levels of table headings for the pairs of columns or combine both UoM and qty into a single field. The latter is what we‘re doing in many currency fields, so that concept should be expandable.

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 1]
```diff
@@ -19,16 +19,25 @@
 
 def execute(filters=None):
 	is_reposting_item_valuation_in_progress()
+	if filters.get("include_uom") and filters.get("type") != "Default UOM":
+		frappe.throw(str("Cannot select filters <b> Include UOM </b> and <b>Select UOM</b> together!"))
```
Comment (rtdany10) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: Translate the string and avoid the HTML tags. Use frappe.bold

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 2]
```diff
@@ -19,16 +19,25 @@
...
 	include_uom = filters.get("include_uom")
+	if filters.get("type") == "Purchase UOM":
+		type = "Purchase UOM"
+	elif filters.get("type") == "Sales UOM":
+		type = "Sales UOM"
+	else:
+		type = ""
```
Comment (rtdany10) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: Unnecessary variable `type`

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 3]
```diff
@@ -363,19 +376,35 @@ def get_item_details(items, sl_entries, include_uom):
...
+		)
+	elif type == "Purchase UOM":
+		query = (
+			query.left_join(ucd)
+			.on((ucd.parent == item.name) & (ucd.uom == item.purchase_uom))
+			.select(ucd.conversion_factor)
+			.as_("purchase_uom_con")
+			.select(ucd.uom)
+			.as_("purchase_uom")
+		)
```
Comment (rtdany10) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: Can be combined, Queries are pretty much the same

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 4]
```diff
@@ -19,16 +19,25 @@
...
 	items = get_items(filters)
 	sl_entries = get_stock_ledger_entries(filters, items)
-	item_details = get_item_details(items, sl_entries, include_uom)
+	item_details = get_item_details(items, sl_entries, include_uom, type)
 	opening_row = get_opening_balance(filters, columns, sl_entries)
 	precision = cint(frappe.db.get_single_value("System Settings", "float_precision"))
 
 	data = []
 	conversion_factors = []
+	type_conversion_factors = []
```
Comment (rtdany10) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: We could use `conversion_factors` itself maybe?

**Review by rtdany10 — COMMENTED** [review]
@Dilshanavs

**Comment by Dilshanavs** [comment]
@rohitwaghchaure

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 5]
```diff
@@ -20,18 +20,16 @@
 def execute(filters=None):
 	is_reposting_item_valuation_in_progress()
 	if filters.get("include_uom") and filters.get("type") != "Default UOM":
-		frappe.throw(str("Cannot select filters <b> Include UOM </b> and <b>Select UOM</b> together!"))
+		frappe.throw(
+			_(
+				f"Cannot select filters {frappe.bold ('Include UOM')} and {frappe.bold('Select UOM')} together!"
+			)
+		)
```
Comment (bosue) [inline erpnext/stock/report/stock_ledger/stock_ledger.py:27]: I think the function calls should be added as separately translated variables, something like this:

```
include_str = _('Include UOM').bold
select_str =  _('Select UOM').bold
frappe.throw(
   _("Cannot select filters {0} and {1} together!".format(include_str, select_str)
)
```

File: erpnext/stock/report/stock_ledger/stock_ledger.js [hunk 6]
```diff
@@ -83,6 +83,13 @@ frappe.query_reports["Stock Ledger"] = {
 			"fieldtype": "Link",
 			"options": "UOM"
 		},
+		{
+			"fieldname": "type",
```
Comment (s-aga-r) [inline erpnext/stock/report/stock_ledger/stock_ledger.js]: `fieldname` should be self-descriptive.

File: erpnext/stock/report/stock_ledger/stock_ledger.js [hunk 7]
```diff
@@ -83,6 +83,13 @@ frappe.query_reports["Stock Ledger"] = {
...
+			"label": __("Select UOM"),
+			"fieldtype": "Select",
+			"options": ["Default UOM","Purchase UOM", "Sales UOM"] ,
```
Comment (s-aga-r) [inline erpnext/stock/report/stock_ledger/stock_ledger.js]: remove space

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 8]
```diff
@@ -19,16 +19,23 @@
 
 def execute(filters=None):
 	is_reposting_item_valuation_in_progress()
+	if filters.get("include_uom") and filters.get("type") != "Default UOM":
+		frappe.throw(
+			_(
+				f"Cannot select filters {frappe.bold ('Include UOM')} and {frappe.bold('Select UOM')} together!"
```
Comment (s-aga-r) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: don't use f-strings in the translation function. Use the `format()` method instead.

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 9]
```diff
@@ -71,8 +77,13 @@ def execute(filters=None):
 
 		if include_uom:
 			conversion_factors.append(item_detail.conversion_factor)
-
-	update_included_uom_in_report(columns, data, include_uom, conversion_factors)
+		if filters.get("type") == "Sales UOM" or filters.get("type") == "Purchase UOM":
```
Comment (s-aga-r) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]:
```suggestion
		if filters.get("type") in ["Sales UOM", "Purchase UOM"]:
```

**Review by s-aga-r — CHANGES_REQUESTED** [review]

**Comment by Dilshanavs** [comment]
@rohitwaghchaure

**Comment by Dilshanavs** [comment]
@s-aga-r

File: erpnext/stock/report/stock_ledger/stock_ledger.py (hunk 5 above)
Comment (rtdany10) [reply erpnext/stock/report/stock_ledger/stock_ledger.py:27]: @Dilshanavs ^

**Comment by rtdany10** [comment]
@Dilshanavs mark comments as resolved if you have made the requested changes. Also run pre-commit.

File: erpnext/stock/report/stock_ledger/stock_ledger.js [hunk 10]
```diff
@@ -83,6 +83,13 @@ frappe.query_reports["Stock Ledger"] = {
 			"fieldtype": "Link",
 			"options": "UOM"
 		},
+		{
+			"fieldname": "select_uom",
+			"label": __("Select UOM"),
+			"fieldtype": "Select",
+			"options": ["Default UOM","Purchase UOM", "Sales UOM"] ,
+			"default": "Default UOM",
```
Comment (s-aga-r) [inline erpnext/stock/report/stock_ledger/stock_ledger.js]: Stock UOM makes more sense IMO.

**Comment by s-aga-r** [comment]
The [test](https://github.com/frappe/erpnext/actions/runs/7068632530/job/19256067558?pr=37387) is failing and seems related.

**Comment by Dilshanavs** [comment]
@s-aga-r

**Comment by Dilshanavs** [comment]
@s-aga-r

**Comment by Dilshanavs** [comment]
@rohitwaghchaure 
@s-aga-r

**Comment by rtdany10** [comment]
@Dilshanavs @rohitwaghchaure @s-aga-r ping

**Pull request closed without merging**